src/librustc/traits/object_safety.rs

```diff
@@ -323,9 +316,108 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             return Some(MethodViolationCode::WhereClauseReferencesSelf(span));
         }
 
+        let receiver_ty = self.liberate_late_bound_regions(
+            method.def_id,
+            &sig.map_bound(|sig| sig.inputs()[0]),
+        );
+
+        // until we get by-value DST, `self: Self` can't be coerced
```

Comment:
In which context do we allow `self: Self` in object-safe traits? Won't this break the "argument types must be sized" requirement?

Reply:
would it be a breaking change to add `where Self: Sized` to  `FnOnce::call_once`? Then I could remove this special-casing until unsized rvalues are implemented

Reply:
No, `FnOnce` is not special. I believe that is the case for all traits. What is disallowed is having a **fn body** (that is, a fn implementation) where a parameter does not have `Sized` type.

Reply:
(Similarly invoking a method is disallowed unless receiver has `Sized` type, I think.)

Reply:
Example:

```rust
trait Foo { fn foo(self); } // ok
trait Bar { fn bar(self) { } } // error: Self: Sized
```

Reply:
The intention was to leave room for "by value DST" later (or some other sol'n).